Spring 5 API Quickstart article (#7973)
auth0/docs: articles/quickstart/backend/java-spring-security5/01-authorization.md

---
title: Authorization
name: How to secure your API using Spring Security 5 and Auth0
description: This tutorial demonstrates how to add authorization to a an API using Spring Security 5.
budicon: 500
topics:
    - quickstart
    - backend
    - java
    - spring-security
github:
    path: 01-Authorization-MVC
contentType: tutorial
useCase: quickstart
---

<%= include('../../../_includes/_api_auth_intro') %>

> <%= include('../_includes/_api_create_new') %>

Reviewer: I think preamble should go first, before starting with the directions

Author: I agree, good catch. We should update other API Quickstarts to do the same (looks like they all place the preamble later, like I did). 

I updated this PR to move the preamble earlier.

---
title: Authorization
name: How to secure your API using Spring Security 5 and Auth0
description: This tutorial demonstrates how to add authorization to a an API using Spring Security 5.
budicon: 500
topics:
    - quickstart
    - backend
    - java
    - spring-security
github:
    path: 01-Authorization-MVC
contentType: tutorial
useCase: quickstart
---

<%= include('../../../_includes/_api_auth_intro') %>

<%= include('../_includes/_api_create_new') %>

<%= include('../_includes/_api_auth_preamble') %>

## Configure the Sample Project

::: note
This Quickstart uses Spring MVC. If you are using Spring WebFlux, the steps to secure an API are similar, but some of the implementation details are different. Refer to the [Spring Security WebFlux Sample Code](https://github.com/auth0-samples/auth0-spring-security5-api-sample/tree/master/01-Authorization-WebFlux) to see how integrate Auth0 with your Spring WebFlux API.
:::

> The sample project uses a `/src/main/resources/application.yml` file which configures it to use the correct Auth0 **Domain** and **API Identifier** for your API. If you download the code from this page it will be automatically configured. If you use the example from GitHub, you will need to fill it yourself.

Reviewer 2: ```suggestion
The sample project uses a `/src/main/resources/application.yml` file, which configures it to use the correct Auth0 **Domain** and **API Identifier** for your API. If you download the code from this page it will be automatically configured. If you use the example from GitHub, you will need to fill it in yourself.
```

```yaml
auth0:
  audience: ${apiIdentifier}
spring:
  security:
    oauth2:
        resourceserver:
            jwt:
                issuer-uri: https://${account.namespace}/
```

| Attribute | Description|
| --- | --- |
| `auth0.audience` | The unique identifier for your API. If you are following the steps in this tutorial it would be `https://quickstarts/api`. |

> | `spring.security.oauth2.resourceserver.jwt.issuer-uri` | The issuer URI of the resource server, which will be the value of the `iss` claim in the JWT that Auth0 will issue. Spring Security will use this property to discover the authorization server's public keys and validate the JWT. The value will be your Auth0 domain with a `https://` prefix and a `/` suffix (the trailing slash is important). |

Reviewer 2: ```suggestion
| `spring.security.oauth2.resourceserver.jwt.issuer-uri` | The issuer URI of the resource server, which will be the value of the `iss` claim in the JWT issued by Auth0. Spring Security will use this property to discover the authorization server's public keys and validate the JWT. The value will be your Auth0 domain with a `https://` prefix and a `/` suffix (the trailing slash is important). |
```

```xml
// pom.xml

<parent>
    <groupId>org.springframework.boot</groupId>
    <artifactId>spring-boot-starter-parent</artifactId>
    <version>2.1.6.RELEASE</version>
    <relativePath/>
</parent>

<dependencies>
    <dependency>
        <groupId>org.springframework.boot</groupId>
        <artifactId>spring-boot-starter-security</artifactId>
    </dependency>
    <dependency>
        <groupId>org.springframework.boot</groupId>
        <artifactId>spring-boot-starter-web</artifactId>
    </dependency>
    <dependency>
        <groupId>org.springframework.security.oauth.boot</groupId>
        <artifactId>spring-security-oauth2-autoconfigure</artifactId>
    </dependency>
    <dependency>
        <groupId>org.springframework.security</groupId>
        <artifactId>spring-security-oauth2-resource-server</artifactId>
    </dependency>
    <dependency>
        <groupId>org.springframework.security</groupId>
        <artifactId>spring-security-oauth2-jose</artifactId>
    </dependency>
    <dependency>
        <groupId>org.springframework.security</groupId>
        <artifactId>spring-security-config</artifactId>
    </dependency>
</dependencies>
```

> ### Configure Resource Server using JWT

Reviewer 2: I'd leave off the `using JWT` part here as it implies that you're using an actual JWT to configure a resource server. Subheads can also just be written in sentence-case.

```suggestion
### Configure the resource server
```

To configure the application as a Resource Server and validate the JWTs, create a class that extends `WebSecurityConfigurerAdapter`, add the `@EnableWebSecurity` annotation, and override the `configure` method:

```java
// src/main/java/auth0/example/security/SecurityConfig.java

@EnableWebSecurity
public class SecurityConfig extends WebSecurityConfigurerAdapter {

    @Override
    public void configure(HttpSecurity http) throws Exception {
        http.oauth2ResourceServer().jwt();
    }
}
```

> ### Validate the API Audience

Reviewer 2: I'm not sure "API Audience" is the correct wording, it's more like the "token audience".

To save confusion, I would actually just write this as:

```suggestion
### Validate the audience
```

    @Value("<%= "${auth0.audience}" %>")
    private String audience;

    @Value("<%= "${spring.security.oauth2.resourceserver.jwt.issuer-uri}" %>")
    private String issuer;

    @Bean
    JwtDecoder jwtDecoder() {
        NimbusJwtDecoderJwkSupport jwtDecoder = (NimbusJwtDecoderJwkSupport)
                JwtDecoders.fromOidcIssuerLocation(issuer);

        OAuth2TokenValidator<Jwt> audienceValidator = new AudienceValidator(audience);
        OAuth2TokenValidator<Jwt> withIssuer = JwtValidators.createDefaultWithIssuer(issuer);
        OAuth2TokenValidator<Jwt> withAudience = new DelegatingOAuth2TokenValidator<>(withIssuer, audienceValidator);

        jwtDecoder.setJwtValidator(withAudience);

        return jwtDecoder;
    }
}
```

## Protect API Endpoints

The routes shown below are available for the following requests:

* `GET /api/public`: available for non-authenticated requests
* `GET /api/private`: available for authenticated requests containing a valid JWT Access Token
* `GET /api/private-scoped`: available for authenticated requests containing a valid JWT Access Token with the `read:messages` scope granted

> The example below shows how to secure API methods using the `HttpSecurity` provided in the `configure()` method of the `SecurityConfig` class. Route matchers are used to restrict access based on the level of authorization requred:

Reviewer 2: > The example below shows how to secure API methods using the `HttpSecurity` ^ provided in...

Is it missing a word here (^)?

Author: Updated to include the word `object`: "... using the HttpSecurity object ..."

> The example below shows how to secure API methods using the `HttpSecurity` provided in the `configure()` method of the `SecurityConfig` class. Route matchers are used to restrict access based on the level of authorization requred:

Reviewer 2: Also, typo:

```suggestion
The example below shows how to secure API methods using the `HttpSecurity` provided in the `configure()` method of the `SecurityConfig` class. Route matchers are used to restrict access based on the level of authorization required:
```

Author: Updated to fix typo.

```java
// src/main/java/com/auth0/example/security/SecurityConfig.java

public class SecurityConfig extends WebSecurityConfigurerAdapter {

    @Override
    public void configure(HttpSecurity http) throws Exception {
        http.authorizeRequests()
                .mvcMatchers("/api/public").permitAll()
                .mvcMatchers("/api/private").authenticated()
                .mvcMatchers("api/private-scoped").hasAuthority("SCOPE_read:messages")
                .and()
                .oauth2ResourceServer().jwt();
    }
}
```

::: note

> By default, Spring Security will create a `GrantedAuthority` for each scope in the `scope` claim of the JWT. This is what enables using the `hasAuthority("SCOPE_read:messages")` method to restrict access to a valid JWT that contains the `read:messages` scope.

Reviewer 2: > By default, Spring Security will create a `GrantedAuthority` ^ for each scope..

I feel like it's missing a word here. Is `GrantedAuthority` an instance of something? An Enum? Collection?

Author: `GrantedAuthority` is an interface, so it could be "Spring Security will create an instance of `GrantedAuthority`...". However, that reads a little long and is kind of redundant IMO. For those familiar with Spring and Spring Security, I think it would read normally.

> If your use case requires using a different claim to make authorization decisions, see the [Spring Security Reference Documentation](https://docs.spring.io/spring-security/site/docs/current/reference/htmlsingle/#oauth2resourceserver-authorization-extraction) for detailed instructions.

Reviewer 2: Consider for conciseness:

```suggestion
If your use case requires different claims to make authorization decisions, see the [Spring Security Reference Documentation](https://docs.spring.io/spring-security/site/docs/current/reference/htmlsingle/#oauth2resourceserver-authorization-extraction) for detailed instructions.
```

:::

> ### Create the API Controller

Reviewer 2: ```suggestion
### Create the API controller
```

    }

    @GetMapping(value = "/private")
    public Message privateEndpoint() {
        return new Message("All good. You can see this because you are Authenticated.");
    }

    @GetMapping(value = "/private-scoped")
    public Message privateScopedEndpoint() {
        return new Message("All good. You can see this because you are Authenticated with a Token granted the 'read:messages' scope");
    }
}
```

## Run and Test Your API

To build and run the sample project, run the `bootRun` Gradle task.

Linux or macOS:
```bash
./gradlew bootRun
```

Windows:
```bash
gradlew.cmd bootRun
```

Using a REST client such as Postman or cURL, issue a `GET` request to `http://localhost:3010/api/public`. You should receive the response:

```json
{"message":"All good. You DO NOT need to be authenticated to call /api/public."}
```

Next, issue a `GET` request to `http://localhost:3010/api/private`. You should receive a `401 Unauthorized` response.

To test that your API is properly secured, you can obtain a test token in the Auth0 Dashboard:

1. Go to the **Machine to Machine Applications** tab for the API you created above.
2. Ensure that your API test application is marked as authorized.

> 3. Click the **Test** tab, then **COPY TOKEN**.

Reviewer 2: Would be cool to have a screenshot here that shows this page. Perhaps highlight the **Copy Token** button on it for visibility.

Author: That's good feedback. I will look into adding screenshots to help clarify.

Author: Updated with screenshots.

Issue a `GET` request to the `/api/private` endpoint, this time passing the token you obtained above as an `Authorization` header set to `Bearer YOUR-API-TOKEN-HERE`. You should then see the response:

```json
{"message":"All good. You can see this because you are Authenticated."}
```

Finally, to test that the `/api/private-scoped` is properly protected by the `read:messages` scope, make a `GET` request to the `/api/private-scoped` endpoint using the same token as above. You should see a `403 Forbidden` response, as this token does not possess the `read:messages` scope.

Back in the Auth0 Dashboard:

1. Go to the **Permissions** tab for the API you created above.
2. Add a permission of `read:messages` and provide a description.
3. Go to the **Machine to Machine Applications** tab.
4. Expand your authorized test application, select the `read:messages` scope, then click **UPDATE** and then **CONTINUE**.

> 5. Click the **Test** tab, then **COPY TOKEN**.

Reviewer 2: A screenshot that encompasses these steps would be good here too.

Author: That's good feedback. I will look into adding screenshots to help clarify.

Author: Updated with screenshots.

---
title: Authorization
name: How to secure your API using Spring Security 5 and Auth0
description: This tutorial demonstrates how to add authorization to a an API using Spring Security 5.
budicon: 500
topics:
    - quickstart
    - backend
    - java
    - spring-security
github:
    path: 01-Authorization-MVC
contentType: tutorial
useCase: quickstart
---

<%= include('../../../_includes/_api_auth_intro') %>

<%= include('../_includes/_api_create_new') %>

<%= include('../_includes/_api_auth_preamble') %>

## Configure the Sample Project

::: note

> This Quickstart uses Spring MVC. If you are using Spring WebFlux, the steps to secure an API are similar, but some of the implementation details are different. Refer to the [Spring Security WebFlux Sample Code](https://github.com/auth0-samples/auth0-spring-security5-api-sample/tree/master/01-Authorization-WebFlux) to see how integrate Auth0 with your Spring WebFlux API.

Reviewer: Normally linking to a GH repo file/folder gets outdated rather quickly. If the changes are only on a single class and its annotations (as far as I remember from our conversation). Why not have a separate section, maybe at the end of this document, explaining that snippet?

Small typo 

```suggestion
This Quickstart uses Spring MVC. If you are using Spring WebFlux, the steps to secure an API are similar, but some of the implementation details are different. Refer to the [Spring Security WebFlux Sample Code](https://github.com/auth0-samples/auth0-spring-security5-api-sample/tree/master/01-Authorization-WebFlux) to see how to integrate Auth0 with your Spring WebFlux API.
```

Author: Regarding having a snippet for WebFlux instead of linking to a GH repo, the issue is Spring MVC and WebFlux are entirely different ways of running an application. So the code changes for this sample are relatively minor - the most notable difference is in the `SecurityConfig`, but there are other changes such as dependencies (use `webflux` dependencies), and APIController (return `Mono` or `Flux`, etc.). So if we have a snippet or another section for Spring WebFlux, it would also require updates to other snippets/sections of the article so people don't get the two mixed up.

Another alternative would be to separate into multiple parts - `Authorization (MVC)` and `Authorization (WebFlux)`, and providing separate documents for each. WDYT?

Author: > Another alternative would be to separate into multiple parts - Authorization (MVC) and Authorization (WebFlux), and providing separate documents for each. WDYT?

For context, the reason I did not do this to begin with is that doing so could make the flow of the Quickstart less natural, in that our Quickstarts typically build on each step (e.g., 1) Authorization, 2) Using your API, 3) Troubleshooting). But in this case, MVC and WebFlux are different; one is not a prerequisite for another.

:::

> The sample project uses a `/src/main/resources/application.yml` file, which configures it to use the correct Auth0 **Domain** and **API Identifier** for your API. If you download the code from this page it will be automatically configured. If you use the example from GitHub, you will need to fill it in yourself.

Reviewer: ```suggestion
The sample project uses a `/src/main/resources/application.yml` file, which configures it to use the correct Auth0 **Domain** and **API Identifier** for your API. If you download the code from this page it will be automatically configured. If you clone the example from GitHub, you will need to fill it in yourself.
```

```yaml
auth0:

>   audience: ${apiIdentifier}

Reviewer: this won't get rendered/replaced when logged in. Are you defining a constants anywhere else?

Author: `apiIdentifier` is used by the docs and will do replacement in the code snippet when logged in. The actual example YAML config file in the repo that will do replacement when downloading the example is:

```yaml
debug: true
server:
  port: 3010

auth0:
  # Replace with the API Identifier for your Auth0 API.
  audience: {API_IDENTIFIER}

# The following is standard Spring Security OAuth2 configuration.
spring:
  security:
    oauth2:
      resourceserver:
        jwt:
          # Replace with the domain of your Auth0 tenant.
          # Note the trailing slash is important!
          issuer-uri: https://{DOMAIN}/
```

Reviewer: This is new to me. Thanks for sharing!

spring:
  security:
    oauth2:
        resourceserver:
            jwt:
                issuer-uri: https://${account.namespace}/
```

| Attribute | Description|
| --- | --- |

> | `auth0.audience` | The unique identifier for your API. If you are following the steps in this tutorial it would be `https://quickstarts/api`. |

Reviewer: if you already know the value (you're mentioning it here), why not use that instead in the snippet above?

Author: In the snippet above we use the `apiIdentifier` value which will be replaced with the value for the chosen API in the dropdown widget in the code snippet viewer. So it shouldn't be hard-coded in that case, so here it makes sense to call out if following along manually.

Here's a screenshot of the docs rendering the code snippet above:

![image](https://user-images.githubusercontent.com/276225/61669560-1649e180-aca6-11e9-8f3d-03fb31cd99a6.png)

> | `spring.security.oauth2.resourceserver.jwt.issuer-uri` | The issuer URI of the resource server, which will be the value of the `iss` claim in the JWT issued by Auth0. Spring Security will use this property to discover the authorization server's public keys and validate the JWT. The value will be your Auth0 domain with a `https://` prefix and a `/` suffix (the trailing slash is important). |

Reviewer: Suggestion, add signature. Since that's what the pub keys are there for.

```suggestion
| `spring.security.oauth2.resourceserver.jwt.issuer-uri` | The issuer URI of the resource server, which will be the value of the `iss` claim in the JWT issued by Auth0. Spring Security will use this property to discover the authorization server's public keys and validate the JWT signature. The value will be your Auth0 domain with a `https://` prefix and a `/` suffix (the trailing slash is important). |
```

            return OAuth2TokenValidatorResult.success();
        }
        return OAuth2TokenValidatorResult.failure(error);
    }
}
```

Update the `SecurityConfig` class to configure a `JwtDecoder` bean that uses the `AudienceValidator`:

```java
// src/main/java/com/auth0/example/security/SecurityConfig.java

@EnableWebSecurity
public class SecurityConfig extends WebSecurityConfigurerAdapter {

    @Value("<%= "${auth0.audience}" %>")
    private String audience;

    @Value("<%= "${spring.security.oauth2.resourceserver.jwt.issuer-uri}" %>")
    private String issuer;

    @Bean
    JwtDecoder jwtDecoder() {
        NimbusJwtDecoderJwkSupport jwtDecoder = (NimbusJwtDecoderJwkSupport)
                JwtDecoders.fromOidcIssuerLocation(issuer);

        OAuth2TokenValidator<Jwt> audienceValidator = new AudienceValidator(audience);
        OAuth2TokenValidator<Jwt> withIssuer = JwtValidators.createDefaultWithIssuer(issuer);
        OAuth2TokenValidator<Jwt> withAudience = new DelegatingOAuth2TokenValidator<>(withIssuer, audienceValidator);

        jwtDecoder.setJwtValidator(withAudience);

        return jwtDecoder;
    }
}
```

## Protect API Endpoints

> The routes shown below are available for the following requests:

Reviewer: You could re-use this one which is used from other quickstarts as well https://github.com/auth0/docs/blob/00ed91772ecc3a44ec80219383f1f98b05419da3/articles/quickstart/backend/_includes/_api_endpoints.md

* `GET /api/private-scoped`: available for authenticated requests containing a valid JWT Access Token with the `read:messages` scope granted

The example below shows how to secure API methods using the `HttpSecurity` object provided in the `configure()` method of the `SecurityConfig` class. Route matchers are used to restrict access based on the level of authorization required:

```java
// src/main/java/com/auth0/example/security/SecurityConfig.java

public class SecurityConfig extends WebSecurityConfigurerAdapter {

    @Override
    public void configure(HttpSecurity http) throws Exception {
        http.authorizeRequests()
                .mvcMatchers("/api/public").permitAll()
                .mvcMatchers("/api/private").authenticated()
                .mvcMatchers("api/private-scoped").hasAuthority("SCOPE_read:messages")
                .and()
                .oauth2ResourceServer().jwt();
    }
}
```

::: note
By default, Spring Security will create a `GrantedAuthority` for each scope in the `scope` claim of the JWT. This is what enables using the `hasAuthority("SCOPE_read:messages")` method to restrict access to a valid JWT that contains the `read:messages` scope.

> If your use case requires different claims to make authorization decisions, see the [Spring Security Reference Documentation](https://docs.spring.io/spring-security/site/docs/current/reference/htmlsingle/#oauth2resourceserver-authorization-extraction) for detailed instructions.

Reviewer: "detailed instructions"... on what? I'd add something more at the end.

```java
// src/main/java/com/auth0/example/model/Message.java

public class Message {
    private final String message;

    Message(String message) {
        this.message = message;
    }

    public String getMessage() {
        return this.message;
    }
}
```

Create a new class named `APIController` to handle requests to the endpoints:

```java
// src/main/java/com/auth0/example/web/APIController.java

@RestController
@RequestMapping(path = "api", produces = MediaType.APPLICATION_JSON_VALUE)
public class APIController {

    @GetMapping(value = "/public")
    public Message publicEndpoint() {
        return new Message("All good. You DO NOT need to be authenticated to call /api/public.");
    }

    @GetMapping(value = "/private")
    public Message privateEndpoint() {
        return new Message("All good. You can see this because you are Authenticated.");
    }

    @GetMapping(value = "/private-scoped")
    public Message privateScopedEndpoint() {
        return new Message("All good. You can see this because you are Authenticated with a Token granted the 'read:messages' scope");
    }
}
```

## Run and Test Your API

> To build and run the sample project, run the `bootRun` Gradle task.

Reviewer: Given that you mentioned Maven at the start, I'd also explain how to run it with maven, if that's possible for this sample and is no longer than a line or two.

Linux or macOS:
```bash
./gradlew bootRun
```

Windows:
```bash
gradlew.cmd bootRun
```

> Using a REST client such as Postman or cURL, issue a `GET` request to `http://localhost:3010/api/public`. You should receive the response:

Reviewer: All this is also explained on other snippets. Check out the backend includes folder for other things that you could reuse. e.g. for this one, it would be `_api_using.md`

https://github.com/auth0/docs/tree/master/articles/quickstart/backend/_includes

Author: The motivation for including this entire section in the first section of the Quickstart article was based on my own experience getting the Spring Security API sample running. The `_api_using.md` document contains information about obtaining and using access tokens in general, along with information that you mention for calling the different endpoints. Couple issues I had with this were:

* I didn't want to read through additional pages to know how to simply test what I just did.
* The Testing section in the `_api_using.md` document describe how to use the token to call the various endpoints, but not how to get one with the correct scopes.

All that said, there is some duplicate info as you mention. What if we:

* Update the `_api_using.md` with instructions on obtaining a test token with and without the proper permissions.
* Remove this section on testing your API from the "Authorization" page of the Quickstart article.
* Also place the instructions to get a test token in the `README` of the sample repo, for those who go right to the repo.

Some of this is related to two separate sources-of-truth (docs and sample), and I believe it's a goal for 2019 to fix that. But what do you think of the above proposal?

Author: Updated `_api_using.md` to include instructions for obtaining test tokens, and removed the Testing instructions from the Authorization article.

```json
{"message":"All good. You DO NOT need to be authenticated to call /api/public."}
```

Next, issue a `GET` request to `http://localhost:3010/api/private`. You should receive a `401 Unauthorized` response.

To test that your API is properly secured, you can obtain a test token in the Auth0 Dashboard:

1. Go to the **Machine to Machine Applications** tab for the API you created above.
2. Ensure that your API test application is marked as authorized.

> ![Enable Test Application](/media/articles/server-apis/api-enable-test-client.png)

Reviewer: If you migrate to use the includes above (desired!) this image would disappear. Don't forget to remove it from the PR in that case.

Author: Updated `_api_using.md` to include instructions for obtaining test tokens, and removed the Testing instructions from the Authorization article.
